### PR TITLE
Allow live metric visualization for remote runs

### PR DESCRIPTION
**Pull request by Maxusmusti**
Adds new options for live-metric-visualizer:
 - Can now specify a remote host
 - Can select individual hosts for prometheus and pmproxy as well
 - Can now change the ports for prometheus and pmproxy

Adds new option for pcp-graf-visualizer:
 - Use an external redis server instead of built-in

Also cleans up unnecessary environment variables and moves work/defaults to run.py

All three visualizers updated, tested, and working as expected

### CODE REVIEW

File: agent/containers/images/visualizers/run.py [hunk 1]
```diff
@@ -76,12 +76,18 @@
...
+else:
+    url = "http://localhost:"
+
 if metric_type == "live" or metric_type == "prom":
     prom_source_url = f"{graf_base}api/datasources"
     payload = {
         "name": "prometheus",
         "type": "prometheus",
-        "url": "http://localhost:9090",
+        "url": url + "9090",
```
Comment (portante) [inline agent/containers/images/visualizers/run.py:90]: We should consider environment variables for the ports, too.

**Review by portante — CHANGES_REQUESTED** [review]

**Review by portante — CHANGES_REQUESTED** [review]

File: agent/containers/images/visualizers/README.md [hunk 2]
```diff
@@ -7,6 +7,9 @@
 - Directions:
   -   `podman pull pbench/live-metric-visualizer`
   -   `podman run --network host pbench/live-metric-visualizer`
+  -   To visualize a run on a remote host, add `-e HOST=<host>`
```
[the remark this answers is not in the document]
Comment (Maxusmusti) [reply agent/containers/images/visualizers/README.md:10]: The visualizer will only ever need to know the host where the TDS is running, as I'm assuming prometheus/pmproxy will always be on that host. Is this ever not the case?

[the remark this answers is not in the document]
Comment (portante) [reply agent/containers/images/visualizers/README.md:10]: Yes, with containers, once can run a Redis server on host and the Tool Data Sink could be on another host.

[the remark this answers is not in the document]
Comment (Maxusmusti) [reply agent/containers/images/visualizers/README.md:10]: Does this mean that pmproxy or prometheus are ever run on the redis host rather than on the TDS host?

File: agent/containers/images/visualizers/DockerfilePostPCP [hunk 3]
```diff
@@ -2,8 +2,8 @@
 FROM quay.io/pbench/live-metric-visualizer:latest
 
 ENV VIS_TYPE="pcp" \
-    COLLECTOR="/usr/libexec/pcp/bin/pmproxy --foreground --timeseries --port=44322 --redishost=localhost --redisport=6379 --config=/etc/pcp/pmproxy/pmproxy.conf" \
-    PCP_VERSION="5.2.1-1.fc33" \
+    COLLECTOR="/usr/libexec/pcp/bin/pmproxy --foreground --timeseries --port=44566 --redishost=localhost --redisport=6379 --config=/etc/pcp/pmproxy/pmproxy.conf" \
```
[the remark this answers is not in the document]
Comment (Maxusmusti) [reply agent/containers/images/visualizers/DockerfilePostPCP:5]: This script is for the pcp post-run visualizer, so those values should not be changeable (although I guess the PM_PORT could be customizable if we wanted). The redis in reference is a local server within the container to run pmproxy and load the data in a time series format.

File: agent/containers/images/visualizers/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbutenhof) [reply agent/containers/images/visualizers/README.md:10]: I think we should probably focus on two environments:

1. Pbench orchestrates everything; we know exactly where every piece is connected because we put it there, and
2. Pbench is controlled through the APIs, with something external orchestrating... we know where nothing is unless we're told explicitly

The "default" self-orchestrated configuration is easy to manage if we support generalized external orchestration, because we just provide our defaults for every option. But anything that's hard coded may interfere with external orchestration scenarios.

File: agent/containers/images/visualizers/DockerfilePostPCP (hunk 3 above)
[the remark this answers is not in the document]
Comment (portante) [reply agent/containers/images/visualizers/DockerfilePostPCP:5]: Where in the `Dockerfile` does the local Redis server get started as a process?

[the remark this answers is not in the document]
Comment (portante) [reply agent/containers/images/visualizers/DockerfilePostPCP:5]: Okay, I see it now in the `run.py` ... see comments there ...

File: agent/containers/images/visualizers/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (portante) [reply agent/containers/images/visualizers/README.md:10]: If a visualizer requires some sub-system for its operation, we either need to provide it explicitly internal to the container and ensure it is available and running in the container, or we need to have that sub-system provided to us explicitly.

Since we require a Redis instance for the operation of the other parts of the service, we can use the instance provided externally.  If one is not provided externally, then we need to ensure a local one is used.

See the comments in the `DockerfilePostPCP` and `run.py` ...

File: agent/containers/images/visualizers/run.py [hunk 4]
```diff
@@ -76,12 +76,22 @@
 headers["Authorization"] = f"Bearer {token}"
 
 metric_type = os.environ["VIS_TYPE"]
+
```
Comment (portante) [inline agent/containers/images/visualizers/run.py:79]: This comment is meant for line 33 above:  I think we only want to start our own Redis server if an external one is not provided.

**Comment by Maxusmusti** [comment]
Can now also individually specify hosts for prometheus and pmproxy for live metric visualizer.

Can now specify an existing redis instance (host necessary, port optional) to use for pcp post-run visualizer instead of default behavior (launch an instance internally).

Both changes tested and working as expected.

File: agent/containers/images/visualizers/run.py [hunk 5]
```diff
@@ -28,9 +28,15 @@
 # Start the collector (if needed)
 collector_cmd = os.environ["COLLECTOR"]
 if not collector_cmd == "":
-    collector = subprocess.Popen(collector_cmd.split(" "))
+    cmd = collector_cmd.split(" ")
     if os.environ["VIS_TYPE"] == "pcp":
-        subprocess.Popen("redis-server")
+        if not os.environ["REDIS_HOST"] == "":
+            cmd[4] = f"--redishost={os.environ['REDIS_HOST']}"
```
Comment (dbutenhof) [inline agent/containers/images/visualizers/run.py:34]: This assumption about the order of the options seems fragile. It's almost like we really want a dict of option to value pairs so that you can rebuild the value of the `--redishost` option without making this sort of assumption.

For example, instead of a single `COLLECTOR` env, you could have separate variables for each option, which you could replace easily as you build a command from the pieces. Or it wouldn't be all that hard to parse the `COLLECTOR` into a dict (you could even use `ArgumentParser` to do it).

But short of that effort, maybe just a clear comment in the dockerfile that the _order_ of options in the `COLLECTOR` command env is critical and must be retained?

**Comment by Maxusmusti** [comment]
Removed a lot of unnecessary environment variables and moved collector commands and var defaults to run.py

**Review by dbutenhof — DISMISSED** [review]

File: agent/containers/images/visualizers/run.py [hunk 6]
```diff
@@ -75,13 +95,34 @@
 token = json.loads(response.content.decode("utf-8"))["key"]
 headers["Authorization"] = f"Bearer {token}"
 
-metric_type = os.environ["VIS_TYPE"]
+if metric_type == "live":
+    url = (
+        f"http://{os.environ['HOST']}:"
+        if not os.environ["HOST"] == ""
+        else "http://localhost:"
+    )
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/containers/images/visualizers/run.py:103]: I was thinking more along the lines of:

```
if metric_type == "live":
    host = os.environ.get("HOST", "localhost")
    prom_host = os.environ.get("PROM_HOST", host)
    prom_url = f"http://{prom_host}:"
    pm_host = os.environ.get("PM_HOST", host)
    pm_url = f"http://{pm_host}:"
    prom_port = os.environ.get("PROM_PORT", "9090")
    pm_port = os.environ.get("PM_PORT", "44566")
else:
    prom_url = "http://localhost:"
    pm_url = prom_url
    prom_port = "9090"
    pm_port = "44566"
```

**Review by portante — CHANGES_REQUESTED** [review]
More for the exercise of learning "pythonic" methods of handling environment variables ...

**Comment by Maxusmusti** [comment]
> More for the exercise of learning "pythonic" methods of handling environment variables ...

This won't work currently, as the env vars are already defined regardless. The dockerfile also doesn't allow declaration without initialization. Is there a better way to handle these vars?

**Comment by portante** [comment]
> > More for the exercise of learning "pythonic" methods of handling environment variables ...
> 
> This won't work currently, as the env vars are already defined regardless. The dockerfile also doesn't allow declaration without initialization. Is there a better way to handle these vars?

Based on the below documentation, you only need to declare the environment variable if you are going to use it in the Dockerfile.
 
https://docs.docker.com/engine/reference/builder/#environment-replacement

File: agent/containers/images/visualizers/run.py [hunk 7]
```diff
@@ -2,56 +2,78 @@
...
+        "--storage.tsdb.path=/data",
+    ]
+    collector = subprocess.Popen(cmd)
 
 # Start the grafana server
 args = [
     "grafana-server",
     "-homepath",
-    os.environ["GF_PATHS_HOME"],
+    os.environ.get("GF_PATHS_HOME"),
```
Comment (portante) [inline agent/containers/images/visualizers/run.py:74]: What happens if `GF_PATHS_HOME` is not defined?

Comment (portante) [reply agent/containers/images/visualizers/run.py:74]: Is there a default value that will work here?

Comment (portante) [reply agent/containers/images/visualizers/run.py:74]: Seems like we should move those defaults from the `Dockerfile` here as well.

File: agent/containers/images/visualizers/DockerfileLive [hunk 8]
```diff
@@ -19,16 +19,13 @@ ADD combo.json .
 ## THE FOLLOWING LINE WAS PARTIALLY TAKEN FROM https://github.com/grafana/grafana-docker/blob/master/Dockerfile#L7#L13
 ENV PATH=/usr/share/grafana/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
```
Comment (portante) [inline agent/containers/images/visualizers/DockerfileLive:20]: I'd just take all of these out of the Dockerfile.

**Review by portante — CHANGES_REQUESTED** [review]

File: agent/containers/images/visualizers/DockerfileLive (hunk 8 above)
Comment (Maxusmusti) [reply agent/containers/images/visualizers/DockerfileLive:20]: These are not used in run.py though, and are necessary for grafana. Should the env vars be simply defined in run.py instead?

**Comment by Maxusmusti** [comment]
Home path env var not used automatically, requires -homepath arg, so moved vars into run.py

**Review by portante — APPROVED** [review]

**Comment by Maxusmusti** [comment]
Rebased

**Review by portante — DISMISSED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by portante — APPROVED** [review]
